gitCore/jGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java

```diff
@@ -286,6 +300,19 @@ private List<IGitCoreReflogEntry> deriveReflogByRefFullName(String refFullName,
     return remoteBranch;
   }
 
+  private String findFirstPossibleBranchNameToCheck(String branchFullName, String branchShortName) {
+    int firstForwardSlashIndexInBranchShortName = branchShortName.indexOf('/');
+    if (firstForwardSlashIndexInBranchShortName == -1) {
+      return branchFullName;
+    } else {
+      @SuppressWarnings("index:argument") String branchesDirectoryName = branchFullName.substring(0,
```

Comment:
Hmmm looking at the implementation.. I'm not sure whether operating on substrings is the best idea 🤔 it's pretty error-prone and hard to understand TBH. How about switching the the representation of **Vavr List of `/`-separated segments**? You'll then have methods like `take(n)` for taking a prefix with the given number of segments, and `mkString("/")` to join them back before passing to `jgitRepoForMainGitDir.resolve`

Reply:
And the entire impl could be sth like

```java
val segments = List.of(branchFullName.split("/"));
for i = 3..segments.length()-1:
  if jgitRepoForMainGitDir.resolve(segments.take(i).mkString("/")):
    return false
return jgitRepoForMainGitDir.resolve(branchFullName)
```